servershr/ServerQAction.c

```diff
@@ -910,7 +978,8 @@ static int send_all(SOCKET sock, char *msg, int len)
     const int bytes = send(sock, msg + sent, len - sent, MSG_NOSIGNAL);
     if (bytes <= 0)
     {
-      sent = bytes;
+      if (bytes != 0)
+        sent = bytes;
```

Comment:
To reach this line, `send()` must have hit an error and thus returned `-1`.   And thus the inner `if` could be omitted by changing the outer if from `<=` to `<`.

```
if (bytes < 0)
{
   sent = bytes;
   break;
}
```

Reply:
Correction: I made a mistake.  Now realize that the code as written is fine.   That is because there are three cases: -1 = update sent and break, 0 = break, >0 = update sent.

Reply:
yes this was intended. 0 indecates socket closed and hence no more data, return with what we got.
-1 is indicating a socket error. which is somewhat unexpected hence we allow to treat the error.